### PR TITLE
feat(forms): Added emitEvent to AbstractControl methods

### LINKED ISSUE #11286: FormGroup.reset should forward emitEvent to FormControl.setValue
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```
It should be possible to foward `emitEvent` from `FormGroup.reset` to `FormControl.setValue`.
Currently `FormGroup.reset` calls `FormControl.reset` calls `FormControl.setValue`.

In example we have multiple valueChanges listener on different controls in this form, and we don't want to trigger them on reset, currently we added a variable to skip all value changes listener.

**Current behavior**
```ts
let skipper = false;
this._form.get('foo').valueChanges.subscribe((value) => {
	if (!skipper) {
		// to something with control 'bar'
	}
});
this._form.get('bar').valueChanges.subscribe((value) => {
	if (!skipper) {
		// to something with control 'foo'
	}
});
this._form.valueChanges.subscribe((value) => {
	skipper = false;
});

skipper = true; // this is for skipping 'foo' and 'bar' change listener
this._form.reset({ foo: '', bar: '' });
```

**Expected/desired behavior**
```ts
this._form.get('foo').valueChanges.subscribe((value) => {
	// to something with control 'bar'
});
this._form.get('bar').valueChanges.subscribe((value) => {
	// to something with control 'foo'
});

this._form.reset({ foo: '', bar: '' }, { emitEvent: false });
```

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.6
 
* **Browser:** [all] 
 
* **Language:** [TypeScript 1.8]

### LINKED ISSUE #11551: FormGroup setValue / patchValue should have emitEvent:boolean
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[ x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior** 

When calling `myFormGroup.setValue({...})` or (`patchValue({...})`), the `myFormGroup.valueChanges` always gets called. That's not always the desired behavior.

**Expected behavior**

Calling `setValue` or `patchValue` on `FormGroup` should have an `emitEvent` option to prevent the `valueChanges` Observable to be triggered.

**Reproduction of the problem**

---

**What is the motivation / use case for changing the behavior?**

I have some kind of circular update graph, which I can't break that easy:

* Changing the URL `queryParams` updates the `FormGroup`
* The `FormGroup`'s `valueChanges` Observable updates the URL's `queryParams`

(currently the only workaround I have is placing the `FormGroup` inside a child `@Component` and then rely on it's `@Input` that only get's triggered if the object's reference pointer changed. But still this will call `setValue` on the `FormGroup` even if the `FormGroup`'s own `valueChanges` Observable initiated the update circle.)

**Please tell us about your environment:**

OSX El Capitan, VSCode, npm

* **Angular version:** 2.0.0-rc.X

Angular 2.0.0-rc.7

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ] 

all
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.0.2
 
* **Node (for AoT issues):** `node --version` =   
v.6.5.0

This issue is somehow related to this one: https://github.com/angular/angular/issues/11286

### PR DESCRIPTION
**Pull request by fank**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features) **-> See "Other information"**
- [x] Docs have been added / updated (for bug fixes / features) **-> See "Other information"**

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Fixes #11286
Fixes #11551

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
I haven't added test because i checked the spec files for `FormArray`, `FormGroup` and both did not contain the current supported "option" `onlySelf` so, if these aren't tested, there is also no need for `emitEvent`. Same behavior for the docs.

### CODE REVIEW

**Comment by vicb** [comment]
please separate your changes:
- style(forms): ... -> style changes
- fix(forms): ... -> actual fix

**Comment by Fank** [comment]
@vicb https://github.com/angular/angular/pull/11948#issuecomment-250092869 thanks for the info, i haven't see the information about logical separating commits, in https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit . But yea you are right, it makes it much easier to read. Fixed it.

**Comment by kara** [comment]
@Fank Thanks for the PR! Can you please add tests for the methods that have changed?

**Comment by Fank** [comment]
@kara as i wrote above in "Other information", there is current no specs for those cases.
How do should i test it?
And should i added test for `onlySelf` too?

**Comment by kara** [comment]
@Fank I'm not sure how that's relevant. Please add tests for the code you changed. Feel free to add tests for onlySelf as well.

**Comment by Fank** [comment]
@kara sry for the 1. question, it was miss written i mean:
How do the test should looks like? Because i don't how to test it, like https://github.com/angular/angular/blob/master/modules/%40angular/forms/test/form_control_spec.ts#L338 for every method?

**Comment by Fank** [comment]
@kara DONE, i'll do the `onlySelf` in a seperate PR. Added tests for the `emitEvent` change.
Found a test in the FormControl, which i used for every test.

File: modules/@angular/forms/test/form_array_spec.ts [hunk 1]
```diff
@@ -165,6 +165,16 @@ export function main() {
           expect(logger).toEqual(['control1', 'control2', 'array', 'form']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             a.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c2.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.setValue('newValue', {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_array_spec.ts]: The array spec should be testing the FormArray's `setValue` method, so instead call:

```ts
a.setValue(['one', 'two'], {emitEvent: false});
```

File: modules/@angular/forms/test/form_array_spec.ts [hunk 2]
```diff
@@ -263,6 +273,16 @@ export function main() {
           expect(logger).toEqual(['control1', 'array', 'form']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             a.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c2.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.patchValue('newValue', {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_array_spec.ts]: Similarly, this should call `a.patchValue`

File: modules/@angular/forms/test/form_array_spec.ts [hunk 3]
```diff
@@ -457,6 +477,17 @@ export function main() {
...
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             a.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c2.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c3.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.reset({}, {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_array_spec.ts]: `a.reset` here (with an array as the first arg)

File: modules/@angular/forms/test/form_control_spec.ts [hunk 4]
```diff
@@ -549,6 +549,16 @@ export function main() {
           expect(logger).toEqual(['control1', 'group']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             g.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c2.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.reset({}, {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_control_spec.ts]: I'd reset the value to `null` here (first arg), as you're not really testing the value

File: modules/@angular/forms/test/form_group_spec.ts [hunk 5]
```diff
@@ -229,6 +229,15 @@ export function main() {
           expect(logger).toEqual(['control1', 'control2', 'group', 'form']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             g.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.setValue('newValue', {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_group_spec.ts]: As this is the FormGroup spec, you'll want to test the FormGroups' setValue method.  

```ts
g.setValue({'one': 'one', 'two': 'two'}, {emitEvent: false});
```

File: modules/@angular/forms/test/form_group_spec.ts [hunk 6]
```diff
@@ -327,6 +336,15 @@ export function main() {
           expect(logger).toEqual(['control1', 'group', 'form']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             g.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.patchValue('newValue', {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_group_spec.ts]: `g.patchValue`

File: modules/@angular/forms/test/form_group_spec.ts [hunk 7]
```diff
@@ -520,6 +538,15 @@ export function main() {
           expect(logger).toEqual(['control1', 'control2', 'group', 'form']);
         });
 
+        it('should not fire an event when explicitly specified', fakeAsync(() => {
+             form.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             g.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+             c.valueChanges.subscribe((value) => { throw 'Should not happen'; });
+
+             c.reset({}, {emitEvent: false});
```
Comment (kara) [inline modules/@angular/forms/test/form_group_spec.ts]: `g.reset`

**Review by kara — CHANGES_REQUESTED** [review]
Thanks for doing the tests!  The FormArray and FormGroup tests need some updates, then we should be good to go.

**Comment by Fank** [comment]
@kara done

**Review by kara — APPROVED** [review]
LGTM! Thanks again for the PR.

**Pull request merged**